### PR TITLE
Marketplace improvements

### PR DESCRIPTION
**Pull request by tsteur**
fixes https://github.com/piwik/piwik/issues/10695

This update brings many new features:
- New config `multi_server_environment` to specify whether Piwik is running in a multi server environment. If so, we disable some features like installing plugins from the Marketplace directly as it would be only installed on one server instead of all of them. Instead they are able to download. Same for plugin updates. This will be also useful later for things that changes config entries etc
- We are now using composer to check whether a dependency is matching the required version or not, allowing us to use constraints as on https://getcomposer.org/doc/articles/versions.md and very likely fixes couple of bugs.
- Marketplace is now it's own plugin and can be disabled / enabled. It is now fully test with Unit, Integration, System and UI tests where we had no tests before . Config `[General]enable_marketplace=0/1` is now disabled and instead we are using the plugin enabled/activated to disable the Marketplace. The updater respects a setting where the Marketplace is already disabled and it won't  enable the Marketplace in this case
- If someone has a subscription for paid plugins, it is now possible to enter a license key, to install all purchased plugins at once and to see a summary of all existing subscriptions. Note: Those features won't work yet!
- Plugins will be later able to release beta versions of their plugins, and if a beta release channel is enabled, users would receive also beta versions of plugins which is helpful to make plugins compatible with a new Piwik version upfront etc
- We now try to update plugins via HTTPS (if `$coreUpdater->isUpdatingOverHttps()`)

### CODE REVIEW

**Comment by tsteur** [comment]
@sgiehl do you have an idea why `LanguagesManagerTest:: testGetTranslationsForLanguages ` fails in https://travis-ci.org/piwik/piwik/jobs/167524936 ? I understand something seems to be not valid but I don't really get what

File: plugins/CorePluginsAdmin/Menu.php [hunk 1]
```diff
@@ -20,9 +20,18 @@ class Menu extends \Piwik\Plugin\Menu
 {
     private $marketplacePlugins;
```
Comment (mattab) [inline plugins/CorePluginsAdmin/Menu.php]: @tsteur as there are now such nice and complete UI  tests & fixture for the Marketplace, would it maybe be do-able to add a single UI test for this " counter-plugins-to-update displayed-in-menu" feature?

**Comment by mattab** [comment]
Updated the only FAQ that mentioned enable_marketplace: [ How do I configure Piwik on a server without Internet? ](https://piwik.org/faq/plugins/faq_21/)

File: plugins/Marketplace/config/test.php [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+use Interop\Container\ContainerInterface;
+use Piwik\Plugins\Marketplace\tests\Framework\Mock\Consumer as MockConsumer;
+use Piwik\Plugins\Marketplace\LicenseKey;
+use Piwik\Plugins\Marketplace\tests\Framework\Mock\Service as MockService;
+use Piwik\Plugins\Marketplace\Input\PurchaseType;
+
+return array(
+    'MarketplaceEndpoint' => function (ContainerInterface $c) {
+        // if you wonder why this here is configured here again, and the same as in `config.php`,
```
Comment (mattab) [inline plugins/Marketplace/config/test.php]: so thoughtful Thomas :smiley:

**Comment by mattab** [comment]
So much code, so many tests... beautifully hand crafted quality software! 

This **Epic**  pull request :boom: is going to be merged **NOW**! :tada:

**Pull request merged**